bin/reth/src/providers.rs

```diff
@@ -0,0 +1,136 @@
+use std::path::Path;
+
+use reth_rpc::JwtSecret;
+use tracing::info;
+
+use crate::dirs::{JwtSecretPath, PlatformPath};
+
+type BoxError = Box<dyn std::error::Error>;
+
+/// The execution layer and consensus layer clients SHOULD accept a configuration parameter:
+/// jwt-secret, which designates a file containing the hex-encoded 256 bit secret key to be used
+/// for verifying/generating JWT tokens.
+///
+/// If such a parameter is given, but the file cannot be read, or does not contain a hex-encoded
+/// key of 256 bits, the client SHOULD treat this as an error.
+///
+/// If such a parameter is not given, the client SHOULD generate such a token, valid for the
+/// duration of the execution, and SHOULD store the hex-encoded secret as a jwt.hex file on
+/// the filesystem. This file can then be used to provision the counterpart client.
+pub(crate) struct SecretProvider(JwtSecret);
+
+impl AsRef<JwtSecret> for SecretProvider {
+    /// Returns a reference to the inner [`JwtSecret`]
+    fn as_ref(&self) -> &JwtSecret {
+        &self.0
+    }
+}
+
+impl TryFrom<&Option<&Path>> for SecretProvider {
+    type Error = BoxError;
+
+    fn try_from(value: &Option<&Path>) -> Result<Self, Self::Error> {
+        Ok(Self(match value {
+            Some(arg) => Self::try_resolve(arg)?,
+            None => Self::try_create()?,
+        }))
+    }
+}
+
+impl SecretProvider {
+    fn try_resolve(fpath: &Path) -> Result<JwtSecret, BoxError> {
+        let hex = std::fs::read_to_string(fpath)?;
+        let secret = JwtSecret::from_hex(hex)?;
+        info!("Loaded secret {secret:?} from {:?}", fpath.as_os_str());
+        Ok(secret)
+    }
+
+    fn try_create() -> Result<JwtSecret, BoxError> {
+        let default = PlatformPath::<JwtSecretPath>::default();
+        let fpath: &std::path::Path = default.as_ref();
```

Comment:
is there no `AsRef<Path>` on `PlatformPath`?

Reply:
In this case the call to `default()` would panic
https://github.com/paradigmxyz/reth/blob/4df228a87eb4f787131783f7df87f7c7a7e21143/bin/reth/src/dirs.rs#L131-L138